gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+		destroy_neigh_hash_table(&fib[0]->u.neigh);
+	}
+
+	return -1;
+}
+
+static int
+init_fib_tbl(struct gk_config *gk_conf)
+{
+	int i;
+	int ret;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	rte_spinlock_init(&ltbl->lock);
+
+	for (i = 0; i < gk_conf->gk_max_num_fib_entries; i++) {
+		ltbl->fib_tbl[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl[i].num_updated_instances);
+
+		ltbl->fib_tbl6[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl6[i].num_updated_instances);
+	}
+
+	/* Set up the FIB entry for the front network prefixes. */
+	ret = setup_net_prefix_fib(0, &gk_conf->net->front, gk_conf);
+	if (ret < 0)
+		goto out;
+
+	/* Set up the FIB entry for the back network prefixes. */
+	if (gk_conf->net->back_iface_enabled) {
+		ret = setup_net_prefix_fib(1, &gk_conf->net->back, gk_conf);
+		if (ret < 0)
+			goto free_front_fibs;
+	}
+
+	ret = 0;
+	goto out;
+
+free_front_fibs:
+	if (gk_conf->neigh_fib_front != NULL)
+		destroy_neigh_hash_table(&gk_conf->neigh_fib_front->u.neigh);
+	if (gk_conf->neigh6_fib_front != NULL)
+		destroy_neigh_hash_table(&gk_conf->neigh6_fib_front->u.neigh6);
+
+out:
+	return ret;
+}
+
+/*
+ * XXX Only instantiate the LPM tables needed, for example,
+ * there's no need for an IPv6 LPM table in an IPv4-only deployment.
+ */
+int
+setup_gk_lpm(struct gk_config *gk_conf, unsigned int socket_id)
+{
+	int ret;
+	struct rte_lpm_config ipv4_lpm_config;
+	struct rte_lpm6_config ipv6_lpm_config;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	ipv4_lpm_config.max_rules = gk_conf->max_num_ipv4_rules;
+	ipv4_lpm_config.number_tbl8s = gk_conf->num_ipv4_tbl8s;
+	ipv6_lpm_config.max_rules = gk_conf->max_num_ipv6_rules;
+	ipv6_lpm_config.number_tbl8s = gk_conf->num_ipv6_tbl8s;
+
+	/*
+	 * The GK blocks only need to create one single IPv4 LPM table
+	 * on the @socket_id, so the @lcore and @identifier are set to 0.
+	 */
+	ltbl->lpm = init_ipv4_lpm("gk", &ipv4_lpm_config, socket_id, 0, 0);
+	if (ltbl->lpm == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	ltbl->fib_tbl = rte_calloc(NULL, gk_conf->gk_max_num_fib_entries,
+		sizeof(struct gk_fib), 0);
+	if (ltbl->fib_tbl == NULL) {
+		ret = -1;
+		goto free_lpm;
+	}
+
+	/*
+	 * The GK blocks only need to create one single IPv6 LPM table
+	 * on the @socket_id, so the @lcore and @identifier are set to 0.
+	 */
+	ltbl->lpm6 = init_ipv6_lpm("gk", &ipv6_lpm_config, socket_id, 0, 0);
+	if (ltbl->lpm6 == NULL) {
+		ret = -1;
+		goto free_lpm_tbl;
+	}
+
+	ltbl->fib_tbl6 = rte_calloc(NULL, gk_conf->gk_max_num_fib_entries,
+		sizeof(struct gk_fib), 0);
+	if (ltbl->fib_tbl6 == NULL) {
+		ret = -1;
+		goto free_lpm6;
+	}
+
+	ret = init_fib_tbl(gk_conf);
+	if (ret < 0)
+		goto free_lpm_tbl6;
+
+	ret = 0;
+	goto out;
+
+free_lpm_tbl6:
+	rte_free(ltbl->fib_tbl6);
+	ltbl->fib_tbl6 = NULL;
+
+free_lpm6:
+	destroy_ipv6_lpm(ltbl->lpm6);
+
+free_lpm_tbl:
+	rte_free(ltbl->fib_tbl);
+	ltbl->fib_tbl = NULL;
+
+free_lpm:
+	destroy_ipv4_lpm(ltbl->lpm);
+
+out:
+	return ret;
+}
```

Comment:
There's not really any error messages here to describe what goes wrong during failures.